nemo/collections/tts/data/audio_trimming.py

```diff
@@ -0,0 +1,292 @@
+        frame_length: int = 2048,
+        frame_step: int = 512,
+        pad_seconds: float = 0.1,
+        volume_norm: bool = True,
+    ):
+        """Voice activity detection (VAD) based silence trimming.
+
+           Args:
+               model_name: NeMo VAD model to load. Valid configurations can be found with
+                 EncDecClassificationModel.list_available_models()
+               vad_sample_rate: Sample rate used for pretrained VAD model.
+               vad_threshold: Softmax probability [0, 1] of VAD output, above which audio frames will be classified
+                 as speech.
+               device: Device "cpu" or "cuda" to use for running the VAD model.
+               frame_length: Length of audio frames to use when doing speech detection. This does not need to match
+                 the frame_length used any other part of the code or model.
+               frame_step: Stride of audio frames to use when doing speech detection. This does not need to match
+                 the frame_step used any other part of the code or model.
+               pad_seconds: Amount of audio to keep before the detected start of speech and after the end of
+                 speech. Set this to at least 0.1 to avoid cutting off any speech audio, with larger values
+                 being safer but increasing the amount of silence left afterwards.
+               volume_norm: Whether to normalize the volume of audio before doing speech detection.
+        """
+        self.device = device
+        self.vad_model = EncDecClassificationModel.from_pretrained(model_name=model_name).eval().to(self.device)
+        self.vad_sample_rate = vad_sample_rate
+        self.vad_threshold = vad_threshold
+
+        self.frame_threshold = frame_threshold
+        self.frame_length = frame_length
+        self.frame_step = frame_step
+
+        self.pad_seconds = pad_seconds
+        self.volume_norm = volume_norm
+
+    def _detect_speech(self, audio: np.array) -> np.array:
+        # Center-pad the audio
+        audio = np.pad(audio, [self.frame_length // 2, self.frame_length // 2])
+
+        # [num_frames, frame_length]
+        audio_frames = librosa.util.frame(
+            audio, frame_length=self.frame_length, hop_length=self.frame_step
+        ).transpose()
+
+        num_frames = audio_frames.shape[0]
+        # [num_frames, frame_length]
+        audio_signal = torch.tensor(audio_frames, dtype=torch.float32, device=self.device)
+        # [1]
+        audio_signal_len = torch.tensor(num_frames * [self.frame_length], dtype=torch.int32, device=self.device)
+
+        # VAD outputs 2 values for each audio frame with logits indicating the likelihood that
+        # each frame is non-speech or speech, respectively.
+        # [num_frames, 2]
+        log_probs = self.vad_model(input_signal=audio_signal, input_signal_length=audio_signal_len)
+        probs = torch.softmax(log_probs, dim=-1)
+        probs = probs.cpu().detach().numpy()
```

Comment:
Since VAD can run on GPU, transferring to CPU is needed to process it as numpy afterwards. Technically I might be able to transfer it after the next 2 lines.

I will swap the order of the detach() and cpu().